File: tests-cypress/cypress/support/keywords/login.keywords.ts
import { routes } from '../constants/routes'
import type { LoginCredentials } from '../models/auth'
import { headerPage } from '../pages/layout/header-page'
import { loginPage } from '../pages/login/login-page'

export function action_openLoginPage() {
  cy.log('Open the login page')

  loginPage.visit()
}

export function action_loginViaUi(credentials: LoginCredentials) {
  cy.log(`Log in via the UI as user: ${credentials.username}`)

  loginPage.visit()
  loginPage.typeUsername(credentials.username)
  loginPage.typePassword(credentials.password)
  if (credentials.rememberMe) loginPage.enableRememberMe()
  loginPage.submit()
}

export function check_loginShellReady() {
  cy.log('Verify the login form is rendered and interactive')

  loginPage.verifyShellReady()
}

export function check_landedOnDashboard() {
  cy.log('Verify redirect to the dashboard after sign-in')

  cy.location('pathname').should('eq', routes.dashboard)
}

export function check_loginValidationErrors() {
  cy.log('Verify required-field validation messages are shown')

  loginPage.verifyFieldErrors()
}

export function check_loginServerError(message: string) {
  cy.log(`Verify the login server error message: ${message}`)

  loginPage.verifyServerError(message)
}

export function action_submitLoginForm(credentials: LoginCredentials) {
  cy.log(`Submit the already open login form as user: ${credentials.username}`)

  if (credentials.username) loginPage.typeUsername(credentials.username)
  if (credentials.password) loginPage.typePassword(credentials.password)
  loginPage.submit()
}

export function action_submitEmptyLoginForm() {
  cy.log('Submit the login form with both fields empty')

  loginPage.submit()
}

export function action_attemptVisit(path: string) {
  cy.log(`Attempt to open the path directly: ${path}`)

  cy.visit(path)
}

export function action_logout() {
  cy.log('Log out via the header button')

  headerPage.logout()
}

export function check_usernameRequiredError(message: string) {
  cy.log(`Verify the username required error: ${message}`)

  loginPage.verifyUsernameError(message)
}

export function check_passwordRequiredError(message: string) {
  cy.log(`Verify the password required error: ${message}`)

  loginPage.verifyPasswordError(message)
}

export function check_noUsernameError() {
  cy.log('Verify no username validation error is shown')

  loginPage.verifyNoUsernameError()
}

export function check_redirectedToLogin() {
  cy.log('Verify the visitor was redirected to the login page')

  cy.location('pathname').should('eq', routes.login)
}

export function check_landedOn(path: string) {
  cy.log(`Verify the browser landed on: ${path}`)

  cy.location('pathname').should('eq', path)
}

File: tests-selenium/src/test/java/com/coinpulse/selenium/keywords/LoginKeywords.java
package com.coinpulse.selenium.keywords;

import com.coinpulse.selenium.constants.Routes;
import com.coinpulse.selenium.constants.Users;
import com.coinpulse.selenium.core.Config;
import com.coinpulse.selenium.pages.DashboardPage;
import com.coinpulse.selenium.pages.HeaderPage;
import com.coinpulse.selenium.pages.LoginPage;
import io.qameta.allure.Allure;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Login keywords: specs call these, never the page object directly.
 */
public class LoginKeywords {

    private final WebDriver driver;
    private final WebDriverWait wait;
    private final LoginPage loginPage;
    private final DashboardPage dashboardPage;
    private final HeaderPage headerPage;

    public LoginKeywords(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        this.loginPage = new LoginPage(driver);
        this.dashboardPage = new DashboardPage(driver);
        this.headerPage = new HeaderPage(driver);
    }

    public void actionOpenLogin() {
        Allure.step("Open the login page", loginPage::open);
    }

    public void actionLogin(String username, String password) {
        Allure.step("Log in as: " + username, () -> loginPage.loginAs(username, password));
    }

    public void actionLoginViaUi(Users.Credentials credentials) {
        Allure.step("Log in via the UI as user: " + credentials.username(), () -> {
            loginPage.open();
            loginPage.loginAs(credentials.username(), credentials.password());
        });
    }

    public void actionSubmitLoginForm(Users.Credentials credentials) {
        Allure.step("Submit the already open login form as user: " + credentials.username(),
                () -> {
                    if (!credentials.username().isEmpty()) {
                        loginPage.typeUsername(credentials.username());
                    }
                    if (!credentials.password().isEmpty()) {
                        loginPage.typePassword(credentials.password());
                    }
                    loginPage.submit();
                });
    }

    public void actionSubmitEmptyLoginForm() {
        Allure.step("Submit the login form with both fields empty", loginPage::submit);
    }

    public void actionAttemptVisit(String path) {
        Allure.step("Attempt to open the path directly: " + path,
                () -> driver.get(Config.baseUrl() + path));
    }

    public void actionLogout() {
        Allure.step("Log out via the header button", headerPage::logout);
    }

    public void checkLoginShellReady() {
        Allure.step("Verify the login form renders and is interactive",
                loginPage::verifyShellReady);
    }

    public void checkLandedOnDashboard() {
        Allure.step("Verify navigation landed on the dashboard", () -> {
            dashboardPage.awaitLoaded();
            assertThat(driver.getCurrentUrl()).contains(Routes.DASHBOARD);
        });
    }

    public void checkLoginServerError(String message) {
        Allure.step("Verify the login server error message: " + message,
                () -> loginPage.verifyServerError(message));
    }

    public void checkLoginValidationErrors() {
        Allure.step("Verify required-field validation messages are shown",
                loginPage::verifyFieldErrors);
    }

    public void checkUsernameRequiredError(String message) {
        Allure.step("Verify the username required error: " + message,
                () -> loginPage.verifyUsernameError(message));
    }

    public void checkPasswordRequiredError(String message) {
        Allure.step("Verify the password required error: " + message,
                () -> loginPage.verifyPasswordError(message));
    }

    public void checkNoUsernameError() {
        Allure.step("Verify no username validation error is shown",
                loginPage::verifyNoUsernameError);
    }

    public void checkRedirectedToLogin() {
        Allure.step("Verify the visitor was redirected to the login page", () -> {
            loginPage.awaitLoaded();
            wait.until(ExpectedConditions.urlContains(Routes.LOGIN));
        });
    }

    public void checkLandedOn(String path) {
        Allure.step("Verify the browser landed on: " + path,
                () -> wait.until(ExpectedConditions.urlContains(path)));
    }
}
